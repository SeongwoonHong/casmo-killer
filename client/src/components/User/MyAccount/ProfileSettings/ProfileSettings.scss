.for-profile-settings {
    > .user-form-button {
        margin: 15px 15px 15px auto;
    }
}
