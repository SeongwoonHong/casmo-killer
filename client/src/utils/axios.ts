import axios from 'axios';

axios.interceptors.request.use((config) => {
  console.log(config);
  return config;
});

function setTokenToHeader(header, csrfToken) {
  axios.defaults.headers.common[header] = csrfToken;
}

export {
  axios,
  setTokenToHeader,
};
