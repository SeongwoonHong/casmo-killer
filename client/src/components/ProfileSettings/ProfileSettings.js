import React, { Component } from 'react';
import { Prompt } from 'react-router-dom';
import axios from 'axios';

import * as storage from '@sharedUtils/storage';
import {
  validateImg,
  validateEmail,
  validateDisplayName
} from '@sharedUtils/inputValidators';

import FormMessage from '@sharedComponents/FormMessage';

import UserPageContainer from '../UserPageContainer';
import UserInputField from '../UserInputField';

import './ProfileSettings.scss';

class ProfileSettings extends Component {

  constructor(props) {

    super(props);

    this.state = {
      isLoading: false,
      isSuccess: false,
      email: {
        value: props.user.email,
        message: ''
      },
      displayName: {
        value: props.user.displayName,
        message: ''
      },
      avatar: {
        value: props.user.avatar,
        message: ''
      },
      message: '',
      emailSuccessMsg: ''
    };

  }

  componentWillReceiveProps(nextProps) {

    const { email, displayName, avatar } = nextProps.user;

    this.setState({
      email: {
        ...this.state.email,
        value: email
      },
      displayName: {
        ...this.state.displayName,
        value: displayName
      },
      avatar: {
        ...this.state.avatar,
        value: avatar
      }
    });

  }

  onChangeHandler = (e) => {
    this.setState({
      [e.name]: {
        ...this.state[e.name],
        value: e.value
      }
    });
  };

  onImageUpload = (e) => {

    this.setState({
      avatar: {
        ...this.state.avatar,
        isLoading: true
      }
    });

    const reader = new FileReader();
    const image = e.target.files[0];

    let message = '';

    reader.onloadend = () => {

      if (image.size > 5000000) {
        message = 'File is too big.';
      } else if (!validateImg(reader.result)) {
        message = 'File format is not supported.';
      }

      this.setState({
        avatar: {
          value: message.length > 0
            ? this.state.avatar.value
            : reader.result,
          message
        },
        isLoading: false
      });

    };

    reader.readAsDataURL(image);

  };

  onSubmitHandler = async () => {

    this.setState({
      isLoading: true,
      message: '',
      emailSuccessMsg: ''
    });

    const {
      email, displayName, avatar
    } = this.state;

    const errorMsg = {
      forEmail: '',
      forDisplayName: ''
    };

    const userData = new FormData();

    errorMsg.forEmail = await validateEmail(email.value);
    errorMsg.forDisplayName = await validateDisplayName(displayName.value);

    this.setState({
      email: {
        ...this.state.email,
        message: errorMsg.forEmail
      },
      displayName: {
        ...this.state.displayName,
        message: errorMsg.forDisplayName
      }
    });

    if (
      errorMsg.forEmail.length === 0 &&
      errorMsg.forDisplayName.length === 0 &&
      avatar.message.length === 0
    ) {

      userData.append('email', email.value);
      userData.append('displayName', displayName.value);

      if (avatar.value) {
        userData.append('avatar', avatar.value);
      }

      try {

        const { data } = await axios.put('/api/user/update/profile', userData);

        if (data && data.user) {

          this.setState({
            isSuccess: true,
            message: data.message,
            emailSuccessMsg: data.emailSuccessMsg
          });

          const user = await storage.set('ckUser', data.user);
          this.props.onSuccess(user);

        } else {

          this.setState({
            isSuccess: false,
            message: 'Failed to communicate with the server.'
          });

        }

      } catch (error) {
        console.error(error);
        this.setState({
          isSuccess: false,
          message: error.response.data.message
        });
      }

    }

    this.setState({ isLoading: false });

  };

  render() {

    const { user } = this.props;

    const {
      isLoading,
      isSuccess,
      email,
      displayName,
      avatar,
      message,
      emailSuccessMsg
    } = this.state;

    const hasBeenEdited = email.value === user.email
      && displayName.value === user.displayName
      && avatar.value === user.avatar;

    const avatarPreview = (avatarState) => {
      if (avatarState.value) {
        return (
          <img
            className="circle avatar-img"
            alt="user-avatar"
            src={ avatarState.value }
          />
        );
      }
      return <span>No Image</span>;
    };

    return (
      <UserPageContainer
        className="Profile-settings"
        isLoading={ isLoading }
        formTitle="Profile Settings"
        formMsg="Update your email, display name, and profile picture."
        onSubmit={ this.onSubmitHandler }
        disabled={ hasBeenEdited }>

        <Prompt
          when={ !hasBeenEdited }
          message="Are you fucking sure? Your information will be lost." />

        <FormMessage
          message={ message }
          type={ isSuccess ? 'success' : 'error '} />
        <FormMessage message={ emailSuccessMsg } type="warning" />

        <FormMessage message={ email.message } />
        <UserInputField
          name="email"
          onChange={ this.onChangeHandler }
          value={ email.value || '' }
          disabled={ user.strategy !== 'local' } />

        <FormMessage message={ displayName.message } />
        <UserInputField
          name="displayName"
          title="Display Name"
          onChange={ this.onChangeHandler }
          value={ displayName.value || '' } />

        <FormMessage message={ avatar.message } />
        <div className="user-form-fields last-field">
          <label>Profile Picture</label>
          <div className="avatar-preview">
            <div className="avatar-wrapper">
              { avatarPreview(avatar) }
            </div>
            <input
              type="file"
              accept="image/*"
              id="profilePicture"
              onChange={ this.onImageUpload } />
            <label htmlFor="profilePicture">
              <span className="user-form-button">
                Upload New Picture
              </span>
              <span>Max 5mb, JPG, or PNG</span>
            </label>
          </div>
        </div>

      </UserPageContainer>
    );

  }

}

export default ProfileSettings;
