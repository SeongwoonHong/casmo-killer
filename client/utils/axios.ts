import axios from 'axios';

const baseURL = 'http://localhost:9000';

axios.defaults.baseURL = baseURL + '/api';

export { axios };
