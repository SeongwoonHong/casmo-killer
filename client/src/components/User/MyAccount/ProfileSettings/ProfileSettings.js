import React, { Component } from 'react';
import axios from 'axios';
import PlainBtn from 'sharedComponents/PlainBtn';
import LoadingCircle from 'sharedComponents/LoadingCircle';

import inputValidator from 'sharedUtils/inputValidator';

import './ProfileSettings.scss';

class ProfileSettings extends Component {

  constructor(props) {

    super(props);

    this.state = {
      email: {
        value: '',
        message: ''
      },
      username: {
        value: '',
        message: ''
      },
      avatar: {
        value: '',
        message: '',
        isLoading: false
      },
      isLoading: false,
      operationStatus: {
        isSuccess: false,
        message: ''
      }
    };

  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      email: {
        ...this.state.email,
        value: nextProps.user.email
      },
      username: {
        ...this.state.username,
        value: nextProps.user.username
      },
      avatar: {
        ...this.state.avatar,
        value: nextProps.user.avatar
      }
    });
  }

  onErrorHandler = (target, message) => {
    this.setState({
      [target]: {
        ...this.state[target],
        message
      }
    });
  };

  onChangeHandler = (e) => {
    this.setState({
      [e.target.name]: {
        ...this.state[e.target.name],
        value: e.target.value
      }
    });
  };

  onImageUpload = (e) => {

    this.setState({
      avatar: {
        ...this.state.avatar,
        isLoading: true
      }
    });

    const reader = new FileReader();
    const image = e.target.files[0];

    reader.onloadend = () => {
      if (image.size > 5000000) {
        this.onErrorHandler('avatar', 'File is too big.');
        this.setState({
          avatar: {
            ...this.state.avatar,
            isLoading: false
          }
        });
      } else {
        this.setState({
          avatar: {
            isLoading: false,
            value: reader.result,
            message: ''
          }
        });
      }
    };

    reader.readAsDataURL(image);

  };

  onSubmitHandler = async (e) => {

    e.preventDefault();

    this.setState({ isLoading: true });

    const {
      email,
      username,
      avatar
    } = this.state;

    let emailMessage = '';
    let usernameMessage = '';

    // TODO: definitely need to refactor input validations

    if (email.value !== this.props.user.email) {
      if (inputValidator.isEmpty(email.value)) {
        emailMessage = 'Please enter your email.';
      } else if (!inputValidator.isEmail(email.value)) {
        emailMessage = 'The provided email address is not valid.';
      } else {
        try {
          const { data } = await axios.get(`/api/user/validate/email/${email.value}`);
          if (data.isDuplicate) {
            emailMessage = 'The email address is already registered.';
          }
        } catch (error) {
          console.error(error.response.error);
          emailMessage = error.response.data.message;
        }
      }
    }

    if (username.value !== this.props.user.username) {
      if (inputValidator.isEmpty(username.value)) {
        usernameMessage = 'Please enter yxour username.';
      } else if (username.value.length < 4) {
        usernameMessage = 'Username must be more than 4 characters.';
      } else if (username.value.length > 20) {
        usernameMessage = 'Username must be less than 20 characters.';
      } else if (inputValidator.hasWhiteSpace(username.value)) {
        usernameMessage = 'Username cannot have any space between words.';
      } else if (!inputValidator.isUsername(username.value)) {
        usernameMessage = 'Username cannot have special characters.';
      } else {
        try {
          const { data } = await axios.get(`/api/user/validate/username/${username.value}`);
          if (data.isDuplicate) {
            usernameMessage = 'The username is already registered.';
          }
        } catch (error) {
          console.error(error.response.data.error);
          usernameMessage = error.response.data.message;
        }
      }
    }

    if (!emailMessage.length && !usernameMessage.length && !avatar.message.length) {
      try {
        const { data } = await axios.put('/api/user/update/all', {
          email: email.value,
          username: username.value,
          avatar: avatar.value
        });
        this.props.onSuccess(data);
      } catch (error) {
        console.error(error);
      }
      this.setState({ isLoading: false });
    } else {
      this.setState({
        email: {
          ...this.state.email,
          message: emailMessage
        },
        username: {
          ...this.state.username,
          message: usernameMessage
        },
        isLoading: false
      });
    }

  };

  render() {

    const { user } = this.props;

    const {
      email,
      username,
      avatar,
      isLoading
    } = this.state;

    const hasBeenEdited = email.value !== user.email
      || username.value !== user.username
      || avatar.value !== user.avatar;

    const errorMessages = (msg) => {
      if (msg.length) {
        return (
          <div className="submit-messages">
            <p>{ msg }</p>
          </div>
        );
      }
      return null;
    };

    const avatarPreview = (currentUser) => {
      if (currentUser.avatar) {
        return (
          <img
            className="circle avatar-img"
            alt="user-avatar"
            src={ this.state.avatar.value }
          />
        );
      }
      return (
        <div className="avatar-placeholder">
          <span>No Image</span>
        </div>
      );
    };

    return (
      <div className="card">
        <form
          noValidate
          onSubmit={ this.onSubmitHandler }
          className="for-profile">
          <div className="settings-header">
            <h4>Profile Settings</h4>
            <p>Update your email, username, and profile picture.</p>
          </div>
          { errorMessages(email.message) }
          <div className="settings-fields">
            <label htmlFor="email">Email</label>
            <input
              type="email"
              id="email"
              name="email"
              value={ email.value }
              onChange={ this.onChangeHandler }
              disabled={ isLoading } />
            <p>This email is linked to your account.</p>
          </div>
          <div className="divider" />
          { errorMessages(username.message) }
          <div className="settings-fields">
            <label htmlFor="username">Display Name</label>
            <input
              type="text"
              id="username"
              name="username"
              value={ username.value }
              onChange={ this.onChangeHandler }
              disabled={ isLoading } />
            <p>This is the display name for your account.</p>
          </div>
          <div className="divider" />
          { errorMessages(avatar.message) }
          <div className="settings-fields last-field">
            <label>Profile Picture</label>
            <div className="uploader">
              {
                avatar.isLoading
                  ? <LoadingCircle color="teal" />
                  : avatarPreview(this.props.user)
              }
              <input
                type="file"
                accept="image/*"
                id="profilePicture"
                onChange={ this.onImageUpload } />
              <label htmlFor="profilePicture">
                <span>Upload</span>
                <span>Max 5mb, JPG, or PNG</span>
              </label>
            </div>
          </div>
          <div className="divider" />
          <button
            type="submit"
            disabled={ isLoading || !hasBeenEdited }>
            {
              isLoading
                ? <LoadingCircle color="#fff" />
                : 'Save Changes'
            }
          </button>
        </form>
      </div>
    );
  }
}

export default ProfileSettings;
