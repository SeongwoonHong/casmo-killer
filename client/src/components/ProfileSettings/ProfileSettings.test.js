/* eslint react/jsx-boolean-value: 0 */
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import axios from 'axios';

import { validateEmail, validateDisplayName } from '@sharedUtils/inputValidators';
import ProfileSettings from './ProfileSettings';

describe('<ProfileSettings />', () => {

  const userData = {
    email: 'ckboard@gmail.com',
    displayName: 'ckboard',
    avatar: 'http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg'
  };

  let component;
  let mockGet;
  let mockPost;
  let mockPut;

  beforeEach(() => {

    component = shallow(
      <ProfileSettings
        user={ userData } />
    );

    mockGet = sinon.stub(axios, 'get');
    mockPost = sinon.stub(axios, 'post');
    mockPut = sinon.stub(axios, 'put');

  });

  afterEach(() => {
    mockGet.restore();
    mockPost.restore();
    mockPut.restore();
  });

  it('should render the component consistently', () => {
    expect(component).toMatchSnapshot();
  });

  it('should update the component state with the user state', () => {

    const newUserData = {
      email: {
        value: 'ckboardtoronto@gmail.com',
        message: ''
      },
      displayName: {
        value: 'ckboardToronto',
        message: ''
      },
      avatar: userData.avatar
    };

    component.setState(newUserData);

    expect(component.state('email').value).toEqual(newUserData.email.value);
    expect(component.state('displayName').value).toEqual(newUserData.displayName.value);
    expect(component.find('.Profile-settings').props().disabled).toEqual(false);

  });

  it('should reject invalid user information and display the message', async () => {

    const testData = {
      email: 'ckboardgmail',
      displayName: 'ckb',
      avatar: 'http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg'
    };

    const wrapper = shallow(
      <ProfileSettings
        user={ testData } />
    );

    wrapper.instance().onSubmitHandler();

    const emailMsg = await validateEmail(testData.email);
    const dnMsg = await validateDisplayName(testData.displayName);

    expect(wrapper.state('email').message).toEqual(emailMsg);
    expect(wrapper.state('displayName').message).toEqual(dnMsg);

  });

  it('should submit the form and emit onSuccess event', async (done) => {

    const spy = sinon.spy();

    const responseData = {
      data: {
        user: {
          isLoggedIn: true
        },
        message: 'server message',
        emailSuccessMsg: 'server success message',
      }
    };

    const promiseOne = Promise.resolve({ data: { isDuplicate: false } });
    const promiseTwo = Promise.resolve({ data: { isDuplicate: false } });
    const promiseThree = Promise.resolve(responseData);

    const testData = {
      email: 'thisshouldpassthetest@gmail.com',
      displayName: 'passthetest',
      avatar: 'http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg'
    };

    mockGet.callsFake(() => promiseOne);
    mockPost.callsFake(() => promiseTwo);
    mockPut.callsFake(() => promiseThree);

    const wrapper = shallow(
      <ProfileSettings
        user={ testData }
        onSuccess={ spy } />
    );

    wrapper.instance().onSubmitHandler();

    await promiseOne;
    await promiseTwo;
    await promiseThree;

    setImmediate(() => {
      expect(spy.calledWith(responseData.data.user)).toEqual(true);
      done();
    });

  });

});
