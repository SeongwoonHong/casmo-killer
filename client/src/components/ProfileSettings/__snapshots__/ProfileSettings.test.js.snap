// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileSettings /> should render the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProfileSettings
    user={
        Object {
            "avatar": "http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg",
            "displayName": "ckboard",
            "email": "ckboard@gmail.com",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Prompt
          message="Are you fucking sure? Your information will be lost."
          when={false}
/>,
        <FormMessage
          message=""
          type="error "
/>,
        <FormMessage
          message=""
          type="warning"
/>,
        <FormMessage
          message=""
          type="error"
/>,
        <UserInputField
          disabled={true}
          isVisible={true}
          message=""
          name="email"
          onChange={[Function]}
          placeholder=""
          title=""
          type=""
          value="ckboard@gmail.com"
/>,
        <FormMessage
          message=""
          type="error"
/>,
        <UserInputField
          disabled={false}
          isVisible={true}
          message=""
          name="displayName"
          onChange={[Function]}
          placeholder=""
          title="Display Name"
          type=""
          value="ckboard"
/>,
        <FormMessage
          message=""
          type="error"
/>,
        <div
          className="user-form-fields last-field"
>
          <label>
                    Profile Picture
          </label>
          <div
                    className="avatar-preview"
          >
                    <div
                              className="avatar-wrapper"
                    >
                              <img
                                        alt="user-avatar"
                                        className="circle avatar-img"
                                        src="http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg"
                              />
                    </div>
                    <input
                              accept="image/*"
                              id="profilePicture"
                              onChange={[Function]}
                              type="file"
                    />
                    <label
                              htmlFor="profilePicture"
                    >
                              <span
                                        className="user-form-button"
                              >
                                        Upload New Picture
                              </span>
                              <span>
                                        Max 5mb, JPG, or PNG
                              </span>
                    </label>
          </div>
</div>,
      ],
      "className": "Profile-settings",
      "disabled": true,
      "formMsg": "Update your email, display name, and profile picture.",
      "formTitle": "Profile Settings",
      "icon": "",
      "isLoading": false,
      "onSubmit": [Function],
      "title": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "message": "Are you fucking sure? Your information will be lost.",
          "when": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "",
          "type": "error ",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "",
          "type": "warning",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": true,
          "isVisible": true,
          "message": "",
          "name": "email",
          "onChange": [Function],
          "placeholder": "",
          "title": "",
          "type": "",
          "value": "ckboard@gmail.com",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": false,
          "isVisible": true,
          "message": "",
          "name": "displayName",
          "onChange": [Function],
          "placeholder": "",
          "title": "Display Name",
          "type": "",
          "value": "ckboard",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              Profile Picture
</label>,
            <div
              className="avatar-preview"
>
              <div
                            className="avatar-wrapper"
              >
                            <img
                                          alt="user-avatar"
                                          className="circle avatar-img"
                                          src="http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg"
                            />
              </div>
              <input
                            accept="image/*"
                            id="profilePicture"
                            onChange={[Function]}
                            type="file"
              />
              <label
                            htmlFor="profilePicture"
              >
                            <span
                                          className="user-form-button"
                            >
                                          Upload New Picture
                            </span>
                            <span>
                                          Max 5mb, JPG, or PNG
                            </span>
              </label>
</div>,
          ],
          "className": "user-form-fields last-field",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Profile Picture",
            },
            "ref": null,
            "rendered": "Profile Picture",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="avatar-wrapper"
>
                  <img
                                    alt="user-avatar"
                                    className="circle avatar-img"
                                    src="http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg"
                  />
</div>,
                <input
                  accept="image/*"
                  id="profilePicture"
                  onChange={[Function]}
                  type="file"
/>,
                <label
                  htmlFor="profilePicture"
>
                  <span
                                    className="user-form-button"
                  >
                                    Upload New Picture
                  </span>
                  <span>
                                    Max 5mb, JPG, or PNG
                  </span>
</label>,
              ],
              "className": "avatar-preview",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <img
                    alt="user-avatar"
                    className="circle avatar-img"
                    src="http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg"
/>,
                  "className": "avatar-wrapper",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "user-avatar",
                    "className": "circle avatar-img",
                    "src": "http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "accept": "image/*",
                  "id": "profilePicture",
                  "onChange": [Function],
                  "type": "file",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <span
                      className="user-form-button"
>
                      Upload New Picture
</span>,
                    <span>
                      Max 5mb, JPG, or PNG
</span>,
                  ],
                  "htmlFor": "profilePicture",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Upload New Picture",
                      "className": "user-form-button",
                    },
                    "ref": null,
                    "rendered": "Upload New Picture",
                    "type": "span",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Max 5mb, JPG, or PNG",
                    },
                    "ref": null,
                    "rendered": "Max 5mb, JPG, or PNG",
                    "type": "span",
                  },
                ],
                "type": "label",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Prompt
            message="Are you fucking sure? Your information will be lost."
            when={false}
/>,
          <FormMessage
            message=""
            type="error "
/>,
          <FormMessage
            message=""
            type="warning"
/>,
          <FormMessage
            message=""
            type="error"
/>,
          <UserInputField
            disabled={true}
            isVisible={true}
            message=""
            name="email"
            onChange={[Function]}
            placeholder=""
            title=""
            type=""
            value="ckboard@gmail.com"
/>,
          <FormMessage
            message=""
            type="error"
/>,
          <UserInputField
            disabled={false}
            isVisible={true}
            message=""
            name="displayName"
            onChange={[Function]}
            placeholder=""
            title="Display Name"
            type=""
            value="ckboard"
/>,
          <FormMessage
            message=""
            type="error"
/>,
          <div
            className="user-form-fields last-field"
>
            <label>
                        Profile Picture
            </label>
            <div
                        className="avatar-preview"
            >
                        <div
                                    className="avatar-wrapper"
                        >
                                    <img
                                                alt="user-avatar"
                                                className="circle avatar-img"
                                                src="http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg"
                                    />
                        </div>
                        <input
                                    accept="image/*"
                                    id="profilePicture"
                                    onChange={[Function]}
                                    type="file"
                        />
                        <label
                                    htmlFor="profilePicture"
                        >
                                    <span
                                                className="user-form-button"
                                    >
                                                Upload New Picture
                                    </span>
                                    <span>
                                                Max 5mb, JPG, or PNG
                                    </span>
                        </label>
            </div>
</div>,
        ],
        "className": "Profile-settings",
        "disabled": true,
        "formMsg": "Update your email, display name, and profile picture.",
        "formTitle": "Profile Settings",
        "icon": "",
        "isLoading": false,
        "onSubmit": [Function],
        "title": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "message": "Are you fucking sure? Your information will be lost.",
            "when": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "",
            "type": "error ",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "",
            "type": "warning",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": true,
            "isVisible": true,
            "message": "",
            "name": "email",
            "onChange": [Function],
            "placeholder": "",
            "title": "",
            "type": "",
            "value": "ckboard@gmail.com",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": false,
            "isVisible": true,
            "message": "",
            "name": "displayName",
            "onChange": [Function],
            "placeholder": "",
            "title": "Display Name",
            "type": "",
            "value": "ckboard",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label>
                Profile Picture
</label>,
              <div
                className="avatar-preview"
>
                <div
                                className="avatar-wrapper"
                >
                                <img
                                                alt="user-avatar"
                                                className="circle avatar-img"
                                                src="http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg"
                                />
                </div>
                <input
                                accept="image/*"
                                id="profilePicture"
                                onChange={[Function]}
                                type="file"
                />
                <label
                                htmlFor="profilePicture"
                >
                                <span
                                                className="user-form-button"
                                >
                                                Upload New Picture
                                </span>
                                <span>
                                                Max 5mb, JPG, or PNG
                                </span>
                </label>
</div>,
            ],
            "className": "user-form-fields last-field",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Profile Picture",
              },
              "ref": null,
              "rendered": "Profile Picture",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div
                    className="avatar-wrapper"
>
                    <img
                                        alt="user-avatar"
                                        className="circle avatar-img"
                                        src="http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg"
                    />
</div>,
                  <input
                    accept="image/*"
                    id="profilePicture"
                    onChange={[Function]}
                    type="file"
/>,
                  <label
                    htmlFor="profilePicture"
>
                    <span
                                        className="user-form-button"
                    >
                                        Upload New Picture
                    </span>
                    <span>
                                        Max 5mb, JPG, or PNG
                    </span>
</label>,
                ],
                "className": "avatar-preview",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <img
                      alt="user-avatar"
                      className="circle avatar-img"
                      src="http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg"
/>,
                    "className": "avatar-wrapper",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "alt": "user-avatar",
                      "className": "circle avatar-img",
                      "src": "http://i1.kym-cdn.com/photos/images/newsfeed/000/250/007/672.jpg",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "img",
                  },
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "accept": "image/*",
                    "id": "profilePicture",
                    "onChange": [Function],
                    "type": "file",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <span
                        className="user-form-button"
>
                        Upload New Picture
</span>,
                      <span>
                        Max 5mb, JPG, or PNG
</span>,
                    ],
                    "htmlFor": "profilePicture",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Upload New Picture",
                        "className": "user-form-button",
                      },
                      "ref": null,
                      "rendered": "Upload New Picture",
                      "type": "span",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Max 5mb, JPG, or PNG",
                      },
                      "ref": null,
                      "rendered": "Max 5mb, JPG, or PNG",
                      "type": "span",
                    },
                  ],
                  "type": "label",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
